README
luaposix
========

By the [luaposix project][GitHub]

[![travis-ci status](https://secure.travis-ci.org/luaposix/luaposix.png?branche=master)](http://travis-ci.org/luaposix/luaposix/builds)

luaposix is a POSIX binding, including curses, for [Lua] 5.1 and 5.2;
like most libraries it simply binds to C APIs on the underlying
system, so it won't work on non-POSIX systems. However, it does try
to detect the level of POSIX conformance of the underlying system and
bind only available APIs.

luaposix is released under the MIT license, like Lua (see [COPYING];
it's basically the same as the BSD license). There is no warranty.

Please report bugs and make suggestions by opening an issue on the
github tracker.

Installation
------------

The simplest way to install luaposix is with [LuaRocks]. To install the
latest release (recommended):

    luarocks install luaposix

To install current git master (for testing):

    luarocks install https://raw.github.com/luaposix/luaposix/release/luaposix-git-1.rockspec

With Lua 5.1, luaposix requires the bitop library from http://bitop.luajit.org/
(On Lua 5.2 it will work whether bitop is installed or not.)

To install without LuaRocks, check out the sources from the
[repository][GitHub] and run the following commands:

    cd luaposix
    ./bootstrap
    ./configure --prefix=INSTALLATION-ROOT-DIRECTORY
    make all check install

Dependencies are listed in the dependencies entry of the file
`rockspec.conf`. You will also need Autoconf and Automake.

See [INSTALL] for `configure` instructions and `configure --help`
for details of available command-line switches.

Use
---

The library is split into two modules. The basic POSIX APIs are in
`posix` and the curses APIs in `curses`.

HTML documentation can be generated with [LDoc] by running `make doc`
or viewed online at <http://luaposix.github.com/luaposix/>.

The authoritative online POSIX reference is at
<http://www.opengroup.org/onlinepubs/007904875/toc.htm>.

Example code
------------

See the example program `tree.lua`, along with the tests in
`tests-*.lua`.

For a complete application, see the lua branch of [GNU Zile].

Bugs reports & patches
----------------------

Bug reports and patches are most welcome. Please use the github issue
tracker (see URL at top). There is no strict coding style, but please
bear in mind the following points when writing new code:

0. Follow existing code. There are a lot of useful patterns and
   avoided traps there.

1. 8-character indentation using TABs. Not my favourite either, but
   better than reformatting the code and losing much of the ability to
   follow the version control history.

2. No non-POSIX APIs; no platform-specific code. When wrapping APIs
   introduced in POSIX 2001 or later, add an appropriate #if. If your
   platform isn't quite POSIX, you may find a gnulib module to bridge
   the gap. If absolutely necessary, use autoconf feature tests.

3. Thin wrappers: although some existing code contradicts this, wrap
   POSIX APIs in the simplest way possible. If necessary, more
   convenient wrappers can be added in Lua (posix.lua).


[Lua]: http://www.lua.org/
[GitHub]: https://github.com/luaposix/luaposix
[LuaRocks]: http://www.luarocksporg "Lua package manager"
[LDoc]: https://github.com/stevedonovan/LDoc "Lua documentation generator"
[COPYING]: https://raw.github.com/luaposix/luaposix/release/COPYING
[INSTALL]: https://raw.github.com/luaposix/luaposix/release/INSTALL
[GNU Zile]: http://git.savannah.gnu.org/cgit/zile.git/log/?h=lua "A cut-down Emacs clone"


NEWS
# luaposix NEWS - User visible changes

## Noteworthy changes in release 33.3.1 (2015-03-01) [stable]

### Incompatible Changes

  - The briefly lived `posix.system` API has been removed.  It was renamed
    to `posix.spawn` shortly after introduction, and can still be accessed
    by the new symbol name.

### Bugs Fixed

  - `local posix = require "posix"` works again, fixing the regression
    introduced in the last release.  The automated Travis checks have been
    improved to catch this sort of bug in future.


## Noteworthy changes in release 33.3.0 (2015-02-28) [stable]

### New Features

  - Improved documentation of `sockaddr` tables for `posix.sys.socket` calls.

  - New `posix.sys.socket.getsockname` binding.

  - Remove the `posix.bit32` library, and use Lua’s built-in `bit32` library.

  - Can now be linked against NetBSD curses, albeit with several functions
    not implemented by that library returning a "not implemented" error as
    a consequence.

  - New functions `popen`, `popen_pipeline` and `pclose` mimic the POSIX
    functions of the same name while allowing tasks to be Lua functions.

  - `fdopen` has been re-added, working on all supported Lua versions.

  - `execx` allows a Lua function or command to be exec’d; `spawn` is now
    implemented in terms of it.

### Incompatible Changes

  - The ncurses-only `KEY_MOUSE` definition has been removed from
    `posix.curses`.

### Bugs Fixed

  - `posix.sys.resource` only provides RLIM_SAVED_CUR and RLIM_SAVED_MAX if
    they are defined by the C library (which FreeBSD 10 does not).

  - `posix.dirent.dir` and `posix.dirent.files` now raise a Lua `error()`
    when unable to open the path argument, for orthogonality with `io.lines`.

  - Workaround for manifest key clash between `posix.so` and `posix.lua` in
    LuaRocks.


## Noteworthy changes in release 33.2.1 (2015-01-04) [stable]

### Bugs Fixed

  - Install posix.curses.html documentation file correctly.


## Noteworthy changes in release 33.2.0 (2015-01-03) [stable]

### New Features

  - QNX support.

### Bugs Fixed

  - `posix.time.gmtime` and `posix.time.localtime` specifications now
    work correctly in January too!


## Noteworthy changes in release 33.1.0 (2014-12-19) [stable]

### New Features

  - New `posix.bit32` module with band, bnot and bor functions that can be
    used from any supported Lua release, without worrying about loading an
    external bit operations library.

  - Preliminary Lua 5.3.0 compatibility.

### Bugs Fixed

  - No more 'Bad Hints' errors from `posix.sys.socket.getaddrinfo` on many
    hosts.

  - `stdlib.setenv` accepts a 3rd argument again.


## Noteworthy changes in release 33.0.0 (2014-11-04) [stable]

### New Features

  - The curses library is fully integrated into luaposix, including reasonably
    comprehensive LDoc documentation (certainly much better than the single
    wooly web-page in previous releases).  For backwards compatibility, all
    APIs are re-exported from the `curses` module.

  - Most constants available through luaposix are now listed in the LDocs of
    the submodule that defines them.

  - For backwards compatibility, everything is still lumped together in the
    `posix.???` namespace, but, now raw APIs have been reorganised according to
    the POSIX header files they belong to:

    ```lua
    local posix  = require "posix"
    local fcntl  = require "posix.fcntl"
    local unistd = require "posix.unistd"

    local fd = fcntl.open ("x",
       bit32.bor (posix.O_WRONLY, posix.O_CREAT), "u=rw")
    unistd.write (fd, "Hello, World!\n")
    unistd.close (fd)
    ```

    This makes the documentation easier to navigate, and for a better mapping
    between luaposix APIs and the C functions they wrap, so translating from
    C is now easier than ever.

  - Each of the newly separated submodules is self-contained, and can be copied
    to another project for compiling and/or loading in a custom Lua runtime. If
    you want to make use of this, in addition to the source for the modules you
    copy, you'll also need at least the files `ext/posix/_helpers.c` and
    `ext/include/lua52compat.c`, and also `ext/include/strlcat.c` for one or
    two of them.

  - Where submodule calls return a table representation of a similar C struct
    from a POSIX API, the Lua return tables have an appropriate `_type` field
    metatable entry compatible with lua-stdlib `std.object.type`.

  - `posix.signal.signal` now accepts the constants `posix.signal.SIG_DFL` or
    `posix.signal.SIG_IGN` in place of the strings "SIG_DFL" and "SIG_IGN".

  - The submodule bindings `posix.time.gmtime`, `posix.time.localtime`,
    `posix.time.mktime`, `posix.time.strftime` and `posix.time.strptime` now
    accept or create PosixTm tables with 1-to-1 field name mappings with the
    POSIX `struct tm`.  The old APIs with custom field names is still available
    as `posix.gmtime`, `posix.localtime`, `posix.mktime`, `posix.strftime` and
    `posix.strptime`.

  - Similarly, `posix.time.nanosleep` now takes and returns a PosixTimespec
    table.  The old API is still available as `posix.nanosleep`.

  - Where supported by the underlying system, `posix.time.clock_getres` and
    `posix.time.clock_gettime` now require a constant inte argument (newly
    defined in the `posix.time` submodule), and returns a PosixTimespec table.
    The old APIs are still available as `posix.clock_getres` and
    `posix.clock_gettime`.

  - Add `posix.unistd.gethostid`. The old `posix.hostid` API is still available
    as an alias.

  - Add group APIs from grp.h: `posix.grp.endgrent`, `posix.grp.getgrent`,
    `posix.grp.getgrgid`, `posix.grp.getgrnam` and `posix.grp.setgrent`.
    Consequently, `posix.getgroup` is now reimplemented in Lua over the POSIX
    APIs.

  - `posix.getgroup` defaults to current effective group when called with no
    arguments, for consistency with `posix.getpasswd` API.


lib/posix.lua
--[[
 POSIX library for Lua 5.1, 5.2 & 5.3.
 (c) Gary V. Vaughan <gary@vaughan.pe>, 2013-2015
 (c) Reuben Thomas <rrt@sc3d.org> 2010-2013
 (c) Natanael Copa <natanael.copa@gmail.com> 2008-2010
]]
--[[--
 Lua POSIX bindings.

 In addition to the convenience functions documented in this module, all
 APIs from submodules are copied into the return table for convenience and
 backwards compatibility.

 @module posix
]]

local bit = require "posix.bit32"
local M = {}


-- For backwards compatibility, copy all table entries into M namespace.
for _, sub in ipairs {
  "ctype", "dirent", "errno", "fcntl", "fnmatch", "getopt", "glob", "grp",
  "libgen", "poll", "pwd", "sched", "signal", "stdio", "stdlib", "sys.msg",
  "sys.resource", "sys.socket", "sys.stat", "sys.statvfs", "sys.time",
  "sys.times", "sys.utsname", "sys.wait", "syslog", "termio", "time",
  "unistd", "utime"
} do
  local t = require ("posix." .. sub)
  for k, v in pairs (t) do
    if k ~= "version" then
      assert(M[k] == nil, "posix namespace clash: " .. sub .. "." .. k)
      M[k] = v
    end
  end
end


-- Inject deprecated APIs (overwriting submodules) for backwards compatibility.
for k, v in pairs (require "posix.deprecated") do
  M[k] = v
end
for k, v in pairs (require "posix.compat") do
  M[k] = v
end

M.version = "posix for " .. _VERSION .. " / luaposix 33.2.1"


local argerror, argtypeerror, checkstring, checktable, toomanyargerror =
  M.argerror, M.argtypeerror, M.checkstring, M.checktable, M.toomanyargerror


-- Code extracted from lua-stdlib with minimal modifications
local list = {
  sub = function (l, from, to)
    local r = {}
    local len = #l
    from = from or 1
    to = to or len
    if from < 0 then
      from = from + len + 1
    end
    if to < 0 then
      to = to + len + 1
    end
    for i = from, to do
      table.insert (r, l[i])
    end
    return r
  end
}
-- end of stdlib code



--- Check permissions like @{posix.unistd.access}, but for euid.
-- Based on the glibc function of the same name. Does not always check
-- for read-only file system, text busy, etc., and does not work with
-- ACLs &c.
-- @function euidaccess
-- @string file file to check
-- @string mode checks to perform (as for access)
-- @return 0 if access allowed; <code>nil</code> otherwise (and errno is set)

local access, set_errno, stat = M.access, M.set_errno, M.stat
local getegid, geteuid, getgid, getuid =
  M.getegid, M.geteuid, M.getgid, M.getuid

local function euidaccess (file, mode)
  local euid, egid = geteuid (), getegid ()

  if getuid () == euid and getgid () == egid then
    -- If we are not set-uid or set-gid, access does the same.
    return access (file, mode)
  end

  local stats = stat (file)
  if not stats then
    return
  end

  -- The super-user can read and write any file, and execute any file
  -- that anyone can execute.
  if euid == 0 and ((not string.match (mode, "x")) or
                    string.match (stats.st_mode, "x")) then
    return 0
  end

  -- Convert to simple list of modes.
  mode = string.gsub (mode, "[^rwx]", "")

  if mode == "" then
    return 0 -- The file exists.
  end

  -- Get the modes we need.
  local granted = stats.st_mode:sub (1, 3)
  if euid == stats.st_uid then
    granted = stats.st_mode:sub (7, 9)
  elseif egid == stats.st_gid or set.new (posix.getgroups ()):member (stats.st_gid) then
    granted = stats.st_mode:sub (4, 6)
  end
  granted = string.gsub (granted, "[^rwx]", "")

  if string.gsub ("[^" .. granted .. "]", mode) == "" then
    return 0
  end
  set_errno (EACCESS)
end

if _DEBUG ~= false then
  M.euidaccess = function (...)
    local argt = {...}
    checkstring ("euidaccess", 1, argt[1])
    checkstring ("euidaccess", 2, argt[2])
    if #argt > 2 then toomanyargerror ("euidaccess", 2, #argt) end
    return euidaccess (...)
  end
else
  M.euidaccess = euidaccess
end


--- Open a pseudo-terminal.
-- Based on the glibc function of the same name.
-- @fixme add support for term and win arguments
-- @treturn[1] int master file descriptor
-- @treturn[1] int slave file descriptor
-- @treturn[1] string slave file name
-- @return[2] nil
-- @treturn[2] string error message

local bit    = require "posix.bit32"
local fcntl  = require "posix.fcntl"
local stdlib = require "posix.stdlib"
local unistd = require "posix.unistd"

local bor = bit.bor
local open, O_RDWR, O_NOCTTY = fcntl.open, fcntl.O_RDWR, fcntl.O_NOCTTY
local grantpt, openpt, ptsname, unlockpt =
  stdlib.grantpt, stdlib.openpt, stdlib.ptsname, stdlib.unlockpt
local close = unistd.close

local function openpty (term, win)
  local ok, errmsg, master, slave, slave_name
  master, errmsg = openpt (bor (O_RDWR, O_NOCTTY))
  if master then
    ok, errmsg = grantpt (master)
    if ok then
      ok, errmsg = unlockpt (master)
      if ok then
	slave_name, errmsg = ptsname (master)
	if slave_name then
          slave, errmsg = open (slave_name, bor (O_RDWR, O_NOCTTY))
	  if slave then
            return master, slave, slave_name
	  end
	end
      end
    end
    close (master)
  end
  return nil, errmsg
end

if _DEBUG ~= false then
  M.openpty = function (...)
    local argt = {...}
    if #argt > 0 then toomanyargerror ("openpty", 0, #argt) end
    return openpty (...)
  end
else
  M.openpty = openpty
end


--- Run a command or function in a sub-process.
-- @function spawn
-- @param task, a string to be executed as a shell command, or a
--   table of arguments to `P.execp` or a Lua function, which should read
--   from standard input, write to standard output, and return an exit code
-- @tparam string ... positional arguments to the program or function
-- @treturn[1] int status exit code, if successful
-- @treturn[1] string exit type, if wait succeeds
-- @treturn[2] nil
-- @treturn[2] string error message

local unpack = table.unpack or unpack -- 5.3 compatibility

local _exit, errno, execp, fork, wait =
  M._exit, M.errno, M.execp, M.fork, M.wait

local function spawn (task, ...)
  local pid, err = fork ()
  if pid == nil then
    return pid, err
  elseif pid == 0 then
    if type (task) == "string" then
      task = {"/bin/sh", "-c", task, ...}
    end
    if type (task) == "table" then
      execp (unpack (task))
      -- Only get here if there's an error; kill the fork
      local _, no = errno ()
      _exit (no)
    else
      _exit (task (...) or 0)
    end
  else
    local _, reason, status = wait (pid)
    return status, reason -- If wait failed, status is nil & reason is error
  end
end

if _DEBUG ~= false then
  M.spawn = function (task, ...)
    local argt, typetask = {task, ...}, type (task)
    if typetask ~= "string" and typetask ~= "table" and typetask ~= "function" then
      argtypeerror ("spawn", 1, "string, table or function", task)
    end
    for i = 2, #argt do
      if type (argt[i]) ~= "string" and type (argt[i]) ~= "nil" then
	argtypeerror ("spawn", i, "string or nil", argt[i])
      end
    end
    return spawn (task, ...)
  end
else
  M.spawn = spawn
end

M.system = M.spawn -- OBSOLETE alias


--- Perform a series of commands and Lua functions as a pipeline.
-- @function pipeline
-- @tparam table t tasks for @{spawn}
-- @func[opt=@{posix.unistd.pipe} pipe_fn function returning a paired read and
--   write file descriptor
-- @treturn int exit code of the chain

local close, dup, dup2, fork, pipe, wait =
  M.close, M.dup, M.dup2, M.fork, M.pipe, M.wait
local STDIN_FILENO, STDOUT_FILENO = M.STDIN_FILENO, M.STDOUT_FILENO

local function pipeline (t, pipe_fn)
  pipe_fn = pipe_fn or pipe

  local pid, read_fd, write_fd, save_stdout
  if #t > 1 then
    read_fd, write_fd = pipe_fn ()
    if not read_fd then
      die ("error opening pipe")
    end
    pid = fork ()
    if pid == nil then
      die ("error forking")
    elseif pid == 0 then -- child process
      if not dup2 (read_fd, STDIN_FILENO) then
        die ("error dup2-ing")
      end
      close (read_fd)
      close (write_fd)
      os.exit (pipeline (list.sub (t, 2), pipe_fn)) -- recurse with remaining arguments
    else -- parent process
      save_stdout = dup (STDOUT_FILENO)
      if not save_stdout then
        die ("error dup-ing")
      end
      if not dup2 (write_fd, STDOUT_FILENO) then
        die ("error dup2-ing")
      end
      close (read_fd)
      close (write_fd)
    end
  end

  local r = spawn (t[1])
  if not r then
    die ("error in fork or wait")
  end
  close (STDOUT_FILENO)

  if #t > 1 then
    close (write_fd)
    wait (pid)
    if not dup2 (save_stdout, STDOUT_FILENO) then
      die ("error dup2-ing")
    end
    close (save_stdout)
  end

  return r
end

if _DEBUG ~= false then
  M.pipeline = function (...)
    local argt = {...}
    checktable ("pipeline", 1, argt[1])
    local typepipefn = type (argt[2])
    if argt[2] ~= nil and type (argt[2]) ~= "function" then
      argtypeerror ("pipeline", 2, "function or nil", argt[2])
    end
    if #argt > 2 then toomanyargerror ("pipeline", 2, #argt) end
    return pipeline (...)
  end
else
  M.pipeline = pipeline
end


--- Perform a series of commands and Lua functions as a pipeline,
-- returning the output of the last stage's <code>stdout</code> as
-- the values of an iterator.
-- @function pipeline_iterator
-- @tparam table t tasks for @{spawn}
-- @func[opt=@{posix.unistd.pipe} pipe_fn function returning a paired read and
--   write file descriptor
-- @return iterator function returning a chunk of output on each call

local close, fork, pipe, read, wait, write =
  M.close, M.fork, M.pipe, M.read, M.wait, M.write
local BUFSIZ, STDIN_FILENO = M.BUFSIZ, M.STDIN_FILENO

function pipeline_iterator (t, pipe_fn)
  local read_fd, write_fd = pipe ()
  if not read_fd then
    die ("error opening pipe")
  end
  table.insert (t, function ()
                     while true do
                       local s = read (STDIN_FILENO, BUFSIZ)
                       if not s or #s == 0 then break end
                       write (write_fd, s)
                     end
                     close (write_fd)
                   end)

  local pid = fork ()
  if pid == nil then
    die ("error forking")
  elseif pid == 0 then -- child process
    os.exit (pipeline (t, pipe_fn))
  else -- parent process
    close (write_fd)
    return function ()
      local s = read (read_fd, BUFSIZ)
      if not s or #s == 0 then
        wait (pid)
        return nil
      end
      return s
    end
  end
end

if _DEBUG ~= false then
  M.pipeline_iterator = function (...)
    local argt = {...}
    checktable ("pipeline_iterator", 1, argt[1])
    local typepipefn = type (argt[2])
    if argt[2] ~= nil and type (argt[2]) ~= "function" then
      argtypeerror ("pipeline_iterator", 2, "function or nil", argt[2])
    end
    if #argt > 2 then toomanyargerror ("pipeline_iterator", 2, #argt) end
    return pipeline_iterator (...)
  end
end


--- Perform a series of commands and Lua functions as a pipeline.
-- Return the output of the last stage's `stdout`.
-- @function pipeline_slurp
-- @tparam table t tasks for @{spawn}
-- @func[opt=@{posix.unistd.pipe} pipe_fn function returning a paired read and
--   write file descriptor
-- @treturn string output of the pipeline

local function pipeline_slurp (t, pipe_fn)
  local out = ""
  for s in pipeline_iterator (t, pipe_fn) do
    out = out .. s
  end
  return out
end

if _DEBUG ~= false then
  M.pipeline_slurp = function (...)
    local argt = {...}
    checktable ("pipeline_slurp", 1, argt[1])
    local typepipefn = type (argt[2])
    if argt[2] ~= nil and type (argt[2]) ~= "function" then
      argtypeerror ("pipeline_slurp", 2, "function or nil", argt[2])
    end
    if #argt > 2 then toomanyargerror ("pipeline_slurp", 2, #argt) end
    return pipeline_slurp (...)
  end
end


--- Add one gettimeofday() returned timeval to another.
-- @function timeradd
-- @param x a timeval
-- @param y another timeval
-- @return x + y, adjusted for usec overflow

local function timeradd (x, y)
  local sec, usec = 0, 0
  if x.sec then sec = sec + x.sec end
  if y.sec then sec = sec + y.sec end
  if x.usec then usec = usec + x.usec end
  if y.usec then usec = usec + y.usec end
  if usec > 1000000 then
    sec = sec + 1
    usec = usec - 1000000
  end

  return { sec = sec, usec = usec }
end

if _DEBUG ~= false then
  M.timeradd = function (...)
    local argt = {...}
    checktable ("timeradd", 1, argt[1])
    checktable ("timeradd", 2, argt[2])
    if #argt > 2 then toomanyargerror ("timeradd", 2, #argt) end
    return timeradd (...)
  end
end


--- Compare one gettimeofday() returned timeval with another
-- @function timercmp
-- @param x a timeval
-- @param y another timeval
-- @return 0 if x and y are equal, >0 if x is newer, <0 if y is newer

local function timercmp (x, y)
  local x = { sec = x.sec or 0, usec = x.usec or 0 }
  local y = { sec = y.sec or 0, usec = y.usec or 0 }
  if x.sec ~= y.sec then
    return x.sec - y.sec
  else
    return x.usec - y.usec
  end
end

if _DEBUG ~= false then
  M.timercmp = function (...)
    local argt = {...}
    checktable ("timercmp", 1, argt[1])
    checktable ("timercmp", 2, argt[2])
    if #argt > 2 then toomanyargerror ("timercmp", 2, #argt) end
    return timercmp (...)
  end
end


--- Subtract one gettimeofday() returned timeval from another.
-- @function timersub
-- @param x a timeval
-- @param y another timeval
-- @return x - y, adjusted for usec underflow

local function timersub (x,y)
  local sec, usec = 0, 0
  if x.sec then sec = x.sec end
  if y.sec then sec = sec - y.sec end
  if x.usec then usec = x.usec end
  if y.usec then usec = usec - y.usec end
  if usec < 0 then
    sec = sec - 1
    usec = usec + 1000000
  end
  return { sec = sec, usec = usec }
end

if _DEBUG ~= false then
  M.timersub = function (...)
    local argt = {...}
    checktable ("timersub", 1, argt[1])
    checktable ("timersub", 2, argt[2])
    if #argt > 2 then toomanyargerror ("timersub", 2, #argt) end
    return timersub (...)
  end
end


return M
